trunk/doc/source/configuration.rst
.. _section_configuration:

Configuration
=============

The Configuration File
----------------------
The parameters that Plumetrack uses for computing fluxes are controlled through the use of a configuration file. The configuration file used for a particular run of the program can be specified using the *-f* option on the command line. For example::
    
    plumetrack -f my_config.cfg example_images

will process all of the images in the example_images folder using the settings stored in the my_config.cfg file. This allows multiple configurations to be stored for different situations (for example for processing images from different cameras).

.. note::
  
  If you do not specify a configuration file using the *-f* option, then Plumetrack will use the default configuration file. This may not be what you want!

The configuration file itself is a JSON file (JavaScript Object Notation). Full details of the JSON format can be found `here <http://json.org/>`_. Plumetrack configuration files consist of a single collection of name/value pairs (equivalent to a dict in Python). A default configuration file will have been created during installation of Plumetrack, and it is recommended to use this as a template for customised configurations. The default configuration file can be found in the My Documents\\plumetrack folder on Windows and ~/.plumetrack folder on Linux systems.


The plumetrack-gui Program
-----------------------------
By far the easiest way to configure Plumetrack is using the graphical user interface (the plumetrack-gui program). See :ref:`section_gui_quickstart` for details of how to use this program. This allows the effects of changing different configuration options to be viewed in realtime. Since there is a certain amount of trial and error involved in creating a successful configuration file this is an important feature! Note that the configuration files created with the GUI can then be used with the command-line version of Plumetrack.
  
As well as the .cfg configuration files, every output file produced by Plumetrack contains a copy of the configuration used to create it. Results files may therefore be used interchangably with configuration files (results files can even be passed to the Plumetrack program as configuration files using the ``-f`` option).



Configuration Options
---------------------
All of the following options *must* be declared in the configuration file, their corresponding inputs in the GUI are shown in the annotated screenshot below. More information about the Farneback algorithm parameters can be found in the `OpenCV documentation <http://docs.opencv.org/trunk/modules/video/doc/motion_analysis_and_object_tracking.html?highlight=farneback#cv2.calcOpticalFlowFarneback>`_.

.. image:: _static/gui_inputs.png
   :width: 400px
   :align: right


.. _date/time format specifiers: https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior

+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| Name                              | GUI Input | Type    | Description                                                                       |
+===================================+===========+=========+===================================================================================+
| **filename_format**               |           | String  | The format of the filenames of the images you want to process (including the file |
|                                   |           |         | extension). The filenames must contain the capture times of the images. The       |
|                                   |           |         | format is specified using the Python `date/time format specifiers`_.              |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **file_extension**                |           | String  | The filename extension used for the image files e.g. `".png"`.                    |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **motion_pix_threshold_low**      |           | Float   | Any pixels that are below this threshold will be masked using random noise and    |
|                                   |           |         | will be excluded from the flux calculation. Set to -1.0 to prevent low-value      |
|                                   |           |         | thresholding. Note that the threshold refers to the pixel values in the raw image.|
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **motion_pix_threshold_high**     |           | Float   | Any pixels that are above this threshold will be masked using random noise and    |
|                                   |           |         | will be excluded from the flux calculation. Set to -1.0 to prevent high-value     |
|                                   |           |         | thresholding. Note that the threshold refers to the pixel values in the raw image.|
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **random_mean**                   |           | Float   | Masked pixels (either due to thresholding or a mask image) are replaced by        |
|                                   |           |         | Gaussian distributed white noise. This parameter controls the mean value of this  |
|                                   |           |         | noise.                                                                            |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **random_sigma**                  |           | Float   | Masked pixels (either due to thresholding or a mask image) are replaced by        |
|                                   |           |         | Gaussian distributed white noise. This parameter controls the standard-deviation  | 
|                                   |           |         | of this  noise.                                                                   |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **mask_image**                    |           | String  | Regions of the image which you do not want to contribute to the flux, or that     |
|                                   |           |         | might interfere with the motion estimation (for example high-contrast boudaries   |
|                                   |           |         | of the volcanic edifice) can be excluded from the calculations by creating a mask |
|                                   |           |         | image. This should be a greyscale image of the same size as the UV images to be   |
|                                   |           |         | processed. Any pixels that are black (pixel value of 0) in the mask image will be |
|                                   |           |         | replaced with random noise and excluded from the flux calculations. Set this to   |
|                                   |           |         | an empty string if you do not wish to use a mask image.                           |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **pixel_size**                    |           | Float   | The size (either the width or the height) of the pixels in the images in meters.  |
|                                   |           |         | Note that this is *not* the area of the pixels.                                   |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **downsizing_factor**             |           | Float   | Computation time scales approximately linearly with the number of pixels in the   |
|                                   |           |         | images. Large speed increases can therefore be achieved by downscaling the images |
|                                   |           |         | prior to computing the flux. Setting the this parameter to a value >1.0 will      |
|                                   |           |         | result in the images being scaled/resized (e.g. if downsizing_factor=2.0, then    |
|                                   |           |         | images will be halved in size) by this amount prior to be processed. This will    |
|                                   |           |         | increase the speed of computation, but will result in coarser flux estimates.     |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **flux_conversion_factor**        |           | Float   | This is used in conjunction with the pixel_size parameter to convert              |
|                                   |           |         | the pixel values in your images into SO2 masses. You need to select a value for   |
|                                   |           |         | this parameter such that `SO2_mass = pixel_value x pixel_size^2 x                 |
|                                   |           |         | flux_conversion_factor`. The fluxes returned will then have units of              |
|                                   |           |         | SO2-mass-units per second.                                                        |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_pyr_scale**           |           | Float   | Farneback algorithm parameter: the image scale (<1.0) to build pyramids for each  |
|                                   |           |         | image; ``farneback_pyr_scale=0.5`` means a classical pyramid, where each next     |
|                                   |           |         | layer is twice smaller than the previous one.                                     |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_levels**              |           | Integer | Farneback algorithm parameter: number of pyramid layers including the initial     |
|                                   |           |         | image; ``farneback_levels=1`` means that no extra layers are created and only the |
|                                   |           |         | original images are used.                                                         |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_winsize**             |           | Integer | Farneback algorithm parameter: averaging window size; larger values increase the  |
|                                   |           |         | algorithm robustness to image noise and give more chances for fast motion         |
|                                   |           |         | detection, but yield more blurred motion field.                                   |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_iterations**          |           | Integer | Farneback algorithm parameter: number of iterations the algorithm does at each    |
|                                   |           |         | pyramid level.                                                                    |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_poly_n**              |           | Integer | Farneback algorithm parameter: size of the pixel neighborhood used to find        |
|                                   |           |         | polynomial expansion in each pixel; larger values mean that the image will be     |
|                                   |           |         | approximated with smoother surfaces, yielding more robust algorithm and more      |
|                                   |           |         | blurred  motion field, typically ``farneback_poly_n`` =5 or 7.                    |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **farneback_poly_sigma**          |           | Float   | Farneback algorithm parameter: standard deviation of the Gaussian that is used to |
|                                   |           |         | smooth derivatives used as a basis for the polynomial expansion; for              |
|                                   |           |         | ``farneback_poly_n=5``, you can set ``farneback_poly_sigma=1.1``, for             |
|                                   |           |         | ``farneback_poly_n=7``, a good value would be ``farneback_poly_sigma=1.5``.       |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **integration_method**            |           | String  | The integration method may be either '1d' or '2d'. The '1d' method computes flux  |
|                                   |           |         | by multiplying each pixel on the the integration line by its corresponding        |
|                                   |           |         | velocity and then integrating along the line's length. The '2d' method considers  |
|                                   |           |         | every pixel in the image and its corresponding velocity to compute which regions  |
|                                   |           |         | of the image cross the integration line between frames. The '1d' method is        |
|                                   |           |         | marginally faster, but the '2d' method is more robust, especially for images      | 
|                                   |           |         | which have large time gap between them, or for highly dynamic plumes.             |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **integration_pix_threshold_low** |           | Float   | Any pixels that are below this threshold will be excluded from the flux           |
|                                   |           |         | calculation. However, it does not change the visibility of the pixels to the      |
|                                   |           |         | motion estimtion algorithm (unlike the thresholds above).                         |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **integration_lines**             |           | List    | List of dicts of integration line parameters, one for each integration line. See  |
|                                   |           |         | table below for details of what integration line parameters must be set.          |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+


Integration line configuration parameters:

+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| Name                              |           | Type    | Description                                                                       |
+===================================+===========+=========+===================================================================================+
| **name**                          |           | String  | A descriptive name to identify the integration line. This will appear as a column |
|                                   |           |         | heading in results files.                                                         |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **integration_points**            |           | List    | A list of [x, y] points (e.g. [[x1, y1], [x2, y2],...] defining the integration   |
|                                   |           |         | line for the flux calculation. x and y are in pixel coordinates - the origin is in|
|                                   |           |         | the top left of n image.                                                          |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+
| **integration_direction**         |           | Integer | Either 1 or -1. Defines which way is positive across the integration line. The    |
|                                   |           |         | positive direction is defined as the cross-product of a unit vector with          |
|                                   |           |         | `integration_direction` sign along the z-axis with a vector along the integration |
|                                   |           |         | line.                                                                             |
+-----------------------------------+-----------+---------+-----------------------------------------------------------------------------------+





trunk/src/plumetrack/settings.py
#Copyright (C) Nial Peters 2014
#
#This file is part of plumetrack.
#
#plumetrack is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
#
#plumetrack is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with plumetrack.  If not, see <http://www.gnu.org/licenses/>.
import os
import sys
import json
from types import FloatType, ListType, IntType, UnicodeType, DictType
from optparse import OptionParser


__default_config = {                  
                    "filename_format": "%Y%m%d_%H%M%S.png",
                    "file_extension": ".png",
                    "custom_image_loader":"",
                    
                    "motion_pix_threshold_low": -1.0,
                    "motion_pix_threshold_high": -1.0,
                    "random_mean": 0.0,
                    "random_sigma":0.0,
                    "mask_image": "", 
                    
                    "pixel_size": 1.0,
                    "flux_conversion_factor": 1.0,
                    "downsizing_factor": 1.0,
                    
                    "farneback_pyr_scale": 0.5,
                    "farneback_levels": 4,
                    "farneback_winsize": 20,
                    "farneback_iterations": 10,
                    "farneback_poly_n": 7,
                    "farneback_poly_sigma": 1.5,
                    
                    "integration_method": "2d",
                    "integration_pix_threshold_low" : -1,
                    "integration_lines": [{
                                           "name":"Integration Line 1",
                                           "integration_points": [[0, 0],[1, 1]],
                                           "integration_direction": 1
                                           }]             
                  
                    }



def is_config_file(filename):
    """
    Returns True if filename looks like a config file (i.e. is a JSON file) and 
    returns False otherwise. This function is used to determine if we are loading
    configs from a results file or a config file.
    """
    
    if not os.path.exists(filename):
        return False
    
    with open(filename,'r') as ifp:
        try:
            json.load(ifp)
        except ValueError:
            return False
    
    return True


def get_default_config():
    """
    Returns a default working config which can be used as a template for 
    customisation.
    """
    return __default_config


def load_config_file(filename):
    """
    Load the configuration file and returns a dictionary of name:value pairs
    that were specified in the file. This function calls validate_config on the 
    configuration before returning it.
    
    Configurations may be loaded either from config files, or from plumetrack
    results files.
    """
    
    if not os.path.exists(filename):
        raise IOError("Failed to open config file \'%s\'. No such file."%filename)
        
    with open(filename,'r') as ifp:
        try:
            config = json.load(ifp)
        except ValueError:
            #maybe we are trying to load a configuration from a results file
            ifp.seek(0)
            config_str = ''
            for line in ifp:
                if line.lstrip().lstrip('#').lstrip().startswith('Configuration'):
                    config_str = line.partition('=')[2]
                    break
            
            try:
                config = json.loads(config_str)
            except ValueError:
                raise IOError("File \"%s\" does not contain a valid plumetrack configuration"%filename)
      
    validate_config(config, filename)
    
    return config



class ConfigError(ValueError):
    """
    Exception raised if the configuration file is invalid.
    """
    pass
     

def __validate_im_loader(config):
    """
    wrapper function used to prevent the setup.py script having to import 
    the image_loader module.
    """
    from plumetrack import image_loader
    return image_loader.validate_loader(config)
     

def validate_config(config, filename=None):
    """
    Raises ConfigError if one or more of the configuration entries is 
    invalid. This function checks that all necessary settings are defined, that 
    they have the correct data type and that they are within sensible limits.
    """
    expected_configs = [
                        ("custom_image_loader",UnicodeType, lambda x: x=="" or __validate_im_loader(config), "Invalid value for \'custom_image_loader\'. Either the file does not exist, is not a valid Python source file, or does not define a subclass of the ImageLoader class."),
                        ("filename_format", UnicodeType, lambda x: config["custom_image_loader"] != "" or (x != "" and not x.isspace()), "\'filename_format\' cannot be an empty string (unless you are using a custom image loader)."),
                        ("file_extension", UnicodeType, lambda x: config["filename_format"]=="" or config["filename_format"].endswith(x), "Mismatch between file extension specified in \'filename_format\' and \'file_extension\'."),
                        ("motion_pix_threshold_low", FloatType, lambda x: (config["motion_pix_threshold_high"] == -1 or x < config["motion_pix_threshold_high"]) and (x == -1 or x >= 0), "\'threshold_low\' must be either -1 or greater than or equal to 0 and must be less than \'threshold_high\'." ),
                        ("motion_pix_threshold_high", FloatType, lambda x: x == -1 or x > 0, "\'threshold_high\' must be either -1 or greater than 0." ),
                        ("random_mean", FloatType, lambda x: True, "" ),
                        ("random_sigma", FloatType, lambda x: True, "" ),
                        ("mask_image", UnicodeType, lambda x: x=="" or x.isspace() or os.path.exists(x), "\'mask_image\' file specified does not exist."),
                        ("pixel_size", FloatType, lambda x: True, "" ),
                        ("flux_conversion_factor", FloatType, lambda x: True, "" ),
                        ("downsizing_factor", FloatType, lambda x: x >=1.0, "\'downsizing_factor\' must be greater than or equal to 1."),
                        ("farneback_pyr_scale", FloatType, lambda x: x<1.0, "\'farneback_pyr_scale\' must be <1.0."),
                        ("farneback_levels", IntType, lambda x: x>=1, "\'farneback_levels\' must be >=1"),
                        ("farneback_winsize", IntType, lambda x: x>=1, "\'farneback_winsize\' must be >=1"),
                        ("farneback_iterations", IntType, lambda x: x>=1, "\'farneback_iterations\' must be >=1"),
                        ("farneback_poly_n", IntType, lambda x: x>=3, "\'farneback_poly_n\' must be >=3 (5 or 7 would be a better choice)"),
                        ("farneback_poly_sigma", FloatType, lambda x: x>0.0, "\'farneback_poly_sigma\' must be >0.0"),
                        ("integration_pix_threshold_low", FloatType, lambda x: (x == -1 or x >= 0), "\'integration_pix_threshold_low\' must be either -1 or >=0."),
                        ("integration_lines", ListType, lambda x: len(x) >= 1, "At least one integration line must be defined."),
                        ("integration_method", UnicodeType, lambda x: x in ('1d','2d'), "\'integration_method\' must be either \'1d\' or \'2d\'")
                        ]
    
    int_line_configs = [
                        ("integration_direction", IntType, lambda x: x==1 or x==-1, "\'integration_direction\' must be either 1 or -1"),
                        ("name", UnicodeType, lambda x: x!="" and not x.isspace(), "Missing name for integration line"),
                        ("integration_points", ListType, lambda x: len(x)>1, "At least two points are required for each integration line.")
                        ]
    
    #first validate all the top-level configs
    __validate_config(config, expected_configs, filename=filename)
    
    #now validate the configs for the integration lines
    for i,c in enumerate(config["integration_lines"]):
        if type(c) != DictType:
            raise ConfigError("Incorrect type for definition of integration "
                              "line %d. Expecting a dict."%(i+1))
        
        __validate_config(c, int_line_configs, filename=filename, 
                          subsection="integration line %d"%(i+1))
    


def __validate_config(config, expected_configs, filename=None, subsection=None):
    #first check that they all exist
    defined_names = config.keys()
    for name in [i[0] for i in expected_configs]:
        try:
            defined_names.remove(name)
        except ValueError:
            if filename is not None:
                if subsection is None:
                    raise ConfigError("Missing definition of %s in configuration file %s"%(name, filename))
                else: 
                    raise ConfigError("Missing definition of %s for %s in configuration file %s"%(name, subsection,filename))
            else:
                if subsection is None:
                    raise ConfigError("Missing definition of %s in configuration."%(name))
                else:
                    raise ConfigError("Missing definition of %s for %s in configuration."%(name, subsection))
                    
    #check if any unknown settings were defined
    if len(defined_names) != 0:
        if filename is not None:
            if subsection is None:
                raise ConfigError("Unknown setting \'%s\' in configuration file \'%s\'."%(defined_names[0], filename))
            else:
                raise ConfigError("Unknown setting \'%s\' for %s in configuration file \'%s\'."%(defined_names[0], subsection, filename))
        else:
            if subsection is None:
                raise ConfigError("Unknown setting \'%s\' in configuration."%(defined_names[0]))
            else:
                raise ConfigError("Unknown setting \'%s\' for %s in configuration."%(defined_names[0], subsection))
            
    #now check all the types and the test_functions
    for name, expected_type, test_func, message in expected_configs:
        x = config[name]
        
        if type(x) != expected_type:
            if expected_type == FloatType:
                #allow floats to be specified as ints and or strings
                try:
                    x = float(x)
                    config[name] = x
                except ValueError:
                    if filename is not None:
                        if subsection is None:
                            raise ConfigError("Incorrect type (%s) for config %s in file %s. Expecting %s"%(type(x),name, filename, expected_type))
                        else:
                            raise ConfigError("Incorrect type (%s) for config %s for %s in file %s. Expecting %s"%(type(x),name, subsection,filename, expected_type))
                    else:
                        if subsection is None:
                            raise ConfigError("Incorrect type (%s) for config %s. Expecting %s"%(type(x),name, expected_type))
                        else:
                            raise ConfigError("Incorrect type (%s) for config %s for %s. Expecting %s"%(type(x),name, subsection, expected_type))
            else:
                if filename is not None:
                    if subsection is None:
                        raise ConfigError("Incorrect type (%s) for config %s in file %s. Expecting %s"%(type(x),name, filename, expected_type))
                    else:
                        raise ConfigError("Incorrect type (%s) for config %s for %s in file %s. Expecting %s"%(type(x),name, subsection,filename, expected_type))

                else:
                    if subsection is None:
                        raise ConfigError("Incorrect type (%s) for config %s. Expecting %s"%(type(x),name, expected_type))
                    else:
                        raise ConfigError("Incorrect type (%s) for config %s for %s. Expecting %s"%(type(x),name, subsection,expected_type))
        if not test_func(x):
            raise ConfigError(message)
    

class OptionError(ValueError):
    pass

class CustomOptionParser(OptionParser):
    def __init__(self, exception_on_error):
        self.exception_on_error = exception_on_error
        OptionParser.__init__(self)
    
    def error(self, message):
        if self.exception_on_error:
            raise OptionError(message)
        else:
            OptionParser.error(self, message)
        


def parse_cmd_line(args=None, exception_on_error=False):
    """
    Function parses the command line input and returns a tuple
    of (options, args). By default this function parses the sys.vargs, but a list
    of command line arguments may be passed in using the args kwarg instead.
    """
    import plumetrack
    usage = ("%prog [options] image_folder")
    
    parser = CustomOptionParser(exception_on_error)
    
    
    parser.prog = plumetrack.PROG_SHORT_NAME
    parser.usage = usage
    parser.description = plumetrack.LONG_DESCRIPTION
    
    parser.add_option("-f", "--config_file", dest="config_file", action='store', 
                      type='string', default=None,
                      help="Specifies the configuration file to use.")
    
    parser.add_option("-r", "--recursive", dest="recursive", 
                      action="store_true", default=False,
                      help="Scan folders recursively for image files")
    
    parser.add_option("-p", "--parallel", dest="parallel", action="store_true", 
                      default=False,
                      help="Parallel process the images. This will be ignored "
                           "when used with the realtime option")
    
    parser.add_option("-t", "--realtime", dest="realtime", action="store_true", 
                      default=False,
                      help="Continuously monitor the image folder for new files")
    
    parser.add_option( "--no_gpu", dest="no_gpu", action="store_true", 
                      default=False,
                      help="Stops plumetrack from using the GPU for processing. "
                      "Note that it may not be using the GPU anyway.")

    parser.add_option("-o","--output_file", dest="output_file", action="store", 
                      default=None, type='string',
                      help="Output file to write computed fluxes to. If no file "
                           "is specified then fluxes will be written to "
                           "standard output. The specified filename may contain "
                           "Python date/time format specifiers to enable "
                           "automatic splitting of output data into multiple "
                           "files. For example a filename of %Y/%m/%d_fluxes.txt "
                           "would split the output into daily files organised by "
                           "month and year.")
    
    parser.add_option("--output_pngs", dest="png_output_folder", action="store", 
                      default=None, type='string',
                      help="Output PNG files of the motion field into this "
                           "folder. The folder (and any subfolders) will be "
                           "created automatically if it does not already exist. "
                           "Default is no PNG output.")
          
    parser.add_option("--max_n", dest="max_n", action="store", 
                      default=None, type='int', help="Sets the maximum number "
                      "of images to process. Note that you will get max_n - 1 "
                      "flux measurements returned (since each flux requires "
                      "two images.)")
                      
    parser.add_option("-s", "--skip_existing", dest="skip_existing", 
                      action="store_true", default=False,
                      help="Ignore images that are already in the folder. Use "
                           "of this option implies --realtime")
    
    parser.add_option("", "--version", action="callback", 
                      callback=__print_version_and_exit, help=("Print Plumetrack"
                      " version number and license information and exit."))

    
    (options, args) = parser.parse_args(args=args)

    if len(args) < 1:
        parser.error("No input folder specified")
    
    elif len(args) > 1:
        parser.error("Only one input folder can be specified. Folders with "
                     "spaces in their names should be enclosed in quotation marks")
    
    elif not os.path.isdir(args[0]):
        parser.error("Cannot open images in \'%s\'. No such directory."%(args[0]))
    
    #a config file must be specified
    if options.config_file is None:
        parser.error("No configuration file specified. You must use the \'-f\' "
                     "option to specify a configuration file to use.")
        
    #use of the skip_existing option implies the use of the realtime option
    if options.skip_existing and not options.realtime:
        options.realtime = True
    
        
    return (options, args)



def __print_version_and_exit(*args):
    """
    Prints the version number and license information for plumetrack.
    """
    import plumetrack #imported here to prevent circular imports
    print "%s (version %s)"%(plumetrack.PROG_SHORT_NAME, plumetrack.VERSION)
    print plumetrack.COPYRIGHT
    print plumetrack.LICENSE_SHORT
    print ""
    print "Written by %s."%plumetrack.AUTHOR
    sys.exit(0)
    